# Assignment is below at the end

- https://scikit-learn.org/stable/modules/tree.html
- https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html
- https://scikit-learn.org/stable/modules/generated/sklearn.tree.plot_tree.html

In [293]:
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
plt.rcParams['figure.figsize'] = (20, 6)
plt.rcParams['font.size'] = 14
import pandas as pd

In [294]:
df = pd.read_csv('adult.data', index_col=False)

In [295]:
df.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  salary  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [296]:
golden = pd.read_csv('adult.test', index_col=False)

In [297]:
golden.head()

age   workclass  fnlwgt      education  education-num       marital-status  \
0   25     Private  226802           11th              7        Never-married   
1   38     Private   89814        HS-grad              9   Married-civ-spouse   
2   28   Local-gov  336951     Assoc-acdm             12   Married-civ-spouse   
3   44     Private  160323   Some-college             10   Married-civ-spouse   
4   18           ?  103497   Some-college             10        Never-married   

           occupation relationship    race      sex  capital-gain  \
0   Machine-op-inspct    Own-child   Black     Male             0   
1     Farming-fishing      Husband   White     Male             0   
2     Protective-serv      Husband   White     Male             0   
3   Machine-op-inspct      Husband   Black     Male          7688   
4                   ?    Own-child   White   Female             0   

   capital-loss  hours-per-week  native-country   salary  
0             0              40   United-States   <=50K.  
1             0              50   United-States   <=50K.  
2             0              40   United-States    >50K.  
3             0              40   United-States    >50K.  
4             0              30   United-States   <=50K.

In [298]:
df.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  salary  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [299]:
df.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'salary'],
      dtype='object')

In [300]:
from sklearn import preprocessing

In [301]:
# Columns we want to transform
transform_columns = ['sex']

#Columns we can't use because non-numerical
non_num_columns = ['workclass', 'education', 'marital-status', 
                     'occupation', 'relationship', 'race', 'sex', 
                     'native-country']

## First let's try using `pandas.get_dummies()` to transform columns

In [302]:
dummies = pd.get_dummies(df[transform_columns])
dummies

sex_ Female  sex_ Male
0                0          1
1                0          1
2                0          1
3                0          1
4                1          0
...            ...        ...
32556            1          0
32557            0          1
32558            1          0
32559            0          1
32560            1          0

[32561 rows x 2 columns]

In [303]:
dummies.shape

(32561, 2)

## sklearn has a similar process for OneHot Encoding features

In [304]:
onehot = preprocessing.OneHotEncoder(handle_unknown="infrequent_if_exist")
onehot.fit(df[transform_columns])

OneHotEncoder(handle_unknown='infrequent_if_exist')

In [305]:
onehot.categories_

[array([' Female', ' Male'], dtype=object)]

In [306]:
sex = onehot.transform(df[transform_columns])
sex

<32561x2 sparse matrix of type '<class 'numpy.float64'>'
	with 32561 stored elements in Compressed Sparse Row format>

In [307]:
sex.shape

(32561, 2)

## In addition to OneHot encoding there is Ordinal Encoding 

In [308]:
enc = preprocessing.OrdinalEncoder()
enc.fit(df[["salary"]])
salary = enc.transform(df[["salary"]])
salary

array([[0.],
       [0.],
       [0.],
       ...,
       [0.],
       [0.],
       [1.]])

In [309]:
enc.categories_[0]

array([' <=50K', ' >50K'], dtype=object)

In [310]:
#x = df.copy()

# transformed = pd.get_dummies(df[transform_columns])
# onehot = preprocessing.OneHotEncoder(handle_unknown="infrequent_if_exist", sparse_output=False).fit(df[transform_columns])
# enc = preprocessing.OrdinalEncoder()
# enc.fit(df[["salary"]])
# transformed = onehot.transform(df[transform_columns])
# new_cols = list(onehot.categories_[0].flatten())
# df_trans = pd.DataFrame(transformed, columns=new_cols)
# x = pd.concat(
    #[
    #    x.drop(non_num_columns, axis=1), 
    #    df_trans
    #], 
    #axis=1,)


#x["salary"] = enc.transform(df[["salary"]])

enc = preprocessing.OrdinalEncoder()
enc.fit(df[["salary"]])

x = df.copy()
x = pd.concat([x.drop(non_num_columns, axis=1),
               pd.get_dummies(df[transform_columns])], axis = 1)
x["salary"] = enc.fit_transform(df[["salary"]])



In [311]:
x.head()

age  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week  \
0   39   77516             13          2174             0              40   
1   50   83311             13             0             0              13   
2   38  215646              9             0             0              40   
3   53  234721              7             0             0              40   
4   28  338409             13             0             0              40   

   salary  sex_ Female  sex_ Male  
0     0.0            0          1  
1     0.0            0          1  
2     0.0            0          1  
3     0.0            0          1  
4     0.0            1          0

In [312]:
xt = golden.copy()

#transformed = onehot.transform(xt[transform_columns])
#new_cols = list(onehot.categories_[0].flatten())
#df_trans = pd.DataFrame(transformed, columns=new_cols)

#x = pd.concat(
  #  [
  #      xt.drop(non_num_columns, axis=1), 
  #      df_trans
   # ], 
  #  axis=1,)

#xt["salary"] = enc.fit_transform(golden[["salary"]])

xt = pd.concat([xt.drop(non_num_columns, axis = 1),
               pd.get_dummies(xt[transform_columns])], axis = 1)
xt["salary"] = enc.fit_transform(xt[["salary"]])

In [313]:
xt.salary.value_counts()

0.0    12435
1.0     3846
Name: salary, dtype: int64

In [314]:
xt.head(5)

age  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week  \
0   25  226802              7             0             0              40   
1   38   89814              9             0             0              50   
2   28  336951             12             0             0              40   
3   44  160323             10          7688             0              40   
4   18  103497             10             0             0              30   

   salary  sex_ Female  sex_ Male  
0     0.0            0          1  
1     0.0            0          1  
2     1.0            0          1  
3     1.0            0          1  
4     0.0            1          0

In [315]:
enc.categories_

[array([' <=50K.', ' >50K.'], dtype=object)]

In [316]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier

#### Choose the model of your preference: DecisionTree or RandomForest

In [317]:
model = RandomForestClassifier(criterion='entropy')

In [318]:
model = DecisionTreeClassifier(criterion='entropy', max_depth=None)

In [319]:
model.fit(x.drop(['fnlwgt','salary'], axis=1), x.salary)

DecisionTreeClassifier(criterion='entropy')

In [320]:
model.tree_.node_count

8337

In [321]:
list(zip(x.drop(['fnlwgt','salary'], axis=1).columns, model.feature_importances_))

[('age', 0.3227419319035386),
 ('education-num', 0.1626704639556689),
 ('capital-gain', 0.22704221349565662),
 ('capital-loss', 0.07906470776873742),
 ('hours-per-week', 0.1529715067592852),
 ('sex_ Female', 0.054127035047242596),
 ('sex_ Male', 0.001382141069870691)]

In [322]:
list(zip(x.drop(['fnlwgt','salary'], axis=1).columns, model.feature_importances_))

[('age', 0.3227419319035386),
 ('education-num', 0.1626704639556689),
 ('capital-gain', 0.22704221349565662),
 ('capital-loss', 0.07906470776873742),
 ('hours-per-week', 0.1529715067592852),
 ('sex_ Female', 0.054127035047242596),
 ('sex_ Male', 0.001382141069870691)]

In [323]:
x.drop(['fnlwgt','salary'], axis=1).head()

age  education-num  capital-gain  capital-loss  hours-per-week  \
0   39             13          2174             0              40   
1   50             13             0             0              13   
2   38              9             0             0              40   
3   53              7             0             0              40   
4   28             13             0             0              40   

   sex_ Female  sex_ Male  
0            0          1  
1            0          1  
2            0          1  
3            0          1  
4            1          0

In [324]:
set(x.columns) - set(xt.columns)

set()

In [325]:
list(x.drop('salary', axis=1).columns)

['age',
 'fnlwgt',
 'education-num',
 'capital-gain',
 'capital-loss',
 'hours-per-week',
 'sex_ Female',
 'sex_ Male']

In [326]:
predictions = model.predict(xt.drop(['fnlwgt','salary'], axis=1))
predictionsx = model.predict(x.drop(['fnlwgt','salary'], axis=1))

In [327]:
from sklearn.metrics import (
    accuracy_score, 
    classification_report, 
    confusion_matrix, auc, roc_curve
)

In [328]:
accuracy_score(xt.salary, predictions)

0.8210797862539156

In [329]:
accuracy_score(xt.salary, predictions)

0.8210797862539156

In [330]:
confusion_matrix(xt.salary, predictions)

array([[11465,   970],
       [ 1943,  1903]], dtype=int64)

In [331]:
print(classification_report(xt.salary, predictions))

              precision    recall  f1-score   support

         0.0       0.86      0.92      0.89     12435
         1.0       0.66      0.49      0.57      3846

    accuracy                           0.82     16281
   macro avg       0.76      0.71      0.73     16281
weighted avg       0.81      0.82      0.81     16281



In [332]:
print(classification_report(xt.salary, predictions))

              precision    recall  f1-score   support

         0.0       0.86      0.92      0.89     12435
         1.0       0.66      0.49      0.57      3846

    accuracy                           0.82     16281
   macro avg       0.76      0.71      0.73     16281
weighted avg       0.81      0.82      0.81     16281



In [333]:
accuracy_score(x.salary, predictionsx)

0.8955806025613464

In [334]:
confusion_matrix(x.salary, predictionsx)

array([[24097,   623],
       [ 2777,  5064]], dtype=int64)

In [335]:
print(classification_report(x.salary, predictionsx))

              precision    recall  f1-score   support

         0.0       0.90      0.97      0.93     24720
         1.0       0.89      0.65      0.75      7841

    accuracy                           0.90     32561
   macro avg       0.89      0.81      0.84     32561
weighted avg       0.90      0.90      0.89     32561



In [336]:
print(classification_report(x.salary, predictionsx))

              precision    recall  f1-score   support

         0.0       0.90      0.97      0.93     24720
         1.0       0.89      0.65      0.75      7841

    accuracy                           0.90     32561
   macro avg       0.89      0.81      0.84     32561
weighted avg       0.90      0.90      0.89     32561



# For the following use the above `adult` dataset. 

# 1. Show the RandomForest outperforms the DecisionTree for a fixed `max_depth` by training using the train set and calculate `precision`, `recall`, `f1`, `confusion matrix` on golden-test set. Start with only numerical features/columns. (age, education-num, capital-gain, capital-loss, hours-per-week) 

In [337]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Selecting numerical features 
num_columns = ['age', 'education-num', 'capital-gain', 'capital-loss', 'hours-per-week']

# Creating training and testing datasets 
x_num = x[num_columns].copy()
xt_num = xt[num_columns].copy()

# Separating features from target
y_train = x['salary']
y_test = xt['salary']

# Training Decision Tree Classifier
dt_model = DecisionTreeClassifier(criterion='entropy', max_depth=15)
dt_model.fit(x_num, y_train)
dt_predictions = dt_model.predict(xt_num)

print("Decision Tree Classifier (Numerical Features Only)")
print("Accuracy:", accuracy_score(y_test, dt_predictions))
print("Confusion Matrix:\n", confusion_matrix(y_test, dt_predictions))
print("Classification Report:\n", classification_report(y_test, dt_predictions))

# Training Random Forest Classifier
rf_model = RandomForestClassifier(criterion='entropy', max_depth=15)
rf_model.fit(x_num, y_train)
rf_predictions = rf_model.predict(xt_num)

print("Random Forest Classifier (Numerical Features Only)")
print("Accuracy:", accuracy_score(y_test, rf_predictions))
print("Confusion Matrix:\n", confusion_matrix(y_test, rf_predictions))
print("Classification Report:\n", classification_report(y_test, rf_predictions))


Decision Tree Classifier (Numerical Features Only)
Accuracy: 0.8325041459369817
Confusion Matrix:
 [[11808   627]
 [ 2100  1746]]
Classification Report:
               precision    recall  f1-score   support

         0.0       0.85      0.95      0.90     12435
         1.0       0.74      0.45      0.56      3846

    accuracy                           0.83     16281
   macro avg       0.79      0.70      0.73     16281
weighted avg       0.82      0.83      0.82     16281

Random Forest Classifier (Numerical Features Only)
Accuracy: 0.839014802530557
Confusion Matrix:
 [[11907   528]
 [ 2093  1753]]
Classification Report:
               precision    recall  f1-score   support

         0.0       0.85      0.96      0.90     12435
         1.0       0.77      0.46      0.57      3846

    accuracy                           0.84     16281
   macro avg       0.81      0.71      0.74     16281
weighted avg       0.83      0.84      0.82     16281



**Interpretation:** Using a fixed max depth of 15, we see that the random forest classifier slightly outperforms the decision tree classifier in terms of accuracy (83.26% vs. 83.76% respectively). In terms of precision, recall, and F1-Score both models produced nearly the same high values on all measures. However, the Random Forest classifier produced slightly better precision (76%) compared to the decision tree classifier (74%). Regarding the confusion matrix, both models had a similar number of true positives and negatives. However, the Random Forest classifier had slightly fewer false positives compared to the Decision Tree classifier. Both models produced similar numbers of false negatives. Overall, the random forest classifier produced marginally better predictions compared to the decision tree classifier. 

# 2. Use a RandomForest or DecisionTree and the `adult` dataset, systematically add new columns, one by one, that are non-numerical but converted using the feature-extraction techniques we learned. Using the golden-test set show [`precision`, `recall`, `f1`, `confusion matrix`] for each additional feature added.

In [338]:
# Pre-processing
drop_columns = ['fnlwgt', 'education', 'occupation', 'relationship', 'native-country']
df2 = df.drop(columns=drop_columns).copy()
golden2 = golden.drop(columns=drop_columns).copy()

# Numeric Columns
num_columns = ['age', 'education-num', 'capital-gain', 'capital-loss', 'hours-per-week']

# Non-Numeric Columns to Encode
non_num_columns = ['workclass', 'marital-status', 'race', 'sex']

# Creating binary variable for salary
df2['salary'] = df2['salary'].map({' <=50K': 0, ' >50K': 1})
golden2['salary'] = golden2['salary'].map({' <=50K.': 0, ' >50K.': 1})

# Encoding non-numeric columns to dummies
df_encoded = pd.get_dummies(df2, columns=non_num_columns)
golden_encoded = pd.get_dummies(golden2, columns=non_num_columns)

# Separating features and target
x = df_encoded.drop(['salary'], axis=1)
xt = golden_encoded.drop(['salary'], axis=1)
y = df_encoded['salary']
yt = golden_encoded['salary']


# Grouping colummns to original features for indexing purposes, n - 1 to avoid dummy variable trap
workclass = ['workclass_ Local-gov', 'workclass_ Never-worked', 'workclass_ Private',
       'workclass_ Self-emp-inc', 'workclass_ Self-emp-not-inc',
       'workclass_ State-gov'] # missing without-pay

marital = ['marital-status_ Divorced', 'marital-status_ Married-AF-spouse',
       'marital-status_ Married-civ-spouse',
       'marital-status_ Married-spouse-absent',
       'marital-status_ Never-married', 'marital-status_ Separated'] # missing widowed

race = ['race_ Amer-Indian-Eskimo',
       'race_ Asian-Pac-Islander', 'race_ Black', 'race_ Other'] # missing White

sex = ['sex_ Female'] # missing Male

In [339]:
# WORKCLASS 

# Creating training and testing datasets with 'workclass' added
x_work = pd.concat([x[num_columns], x[workclass]], axis=1)
xt_work = pd.concat([xt[num_columns], xt[workclass]], axis=1)

# Separating features from target
y_work = y.copy()
yt_work = yt.copy()

# Initialize and train Random Forest Classifier
rf_model = RandomForestClassifier(criterion='entropy', max_depth=15, random_state=42)
rf_model.fit(x_work, y_work)

# Predict on test set
rf_predictions = rf_model.predict(xt_work)

# Evaluate model performance
print("Random Forest Classifier (Numerical Features + 'workclass')")
print("Accuracy:", accuracy_score(yt_work, rf_predictions))
print("Confusion Matrix:\n", confusion_matrix(yt_work, rf_predictions))
print("Classification Report:\n", classification_report(yt_work, rf_predictions))

Random Forest Classifier (Numerical Features + 'workclass')
Accuracy: 0.8383391683557521
Confusion Matrix:
 [[11930   505]
 [ 2127  1719]]
Classification Report:
               precision    recall  f1-score   support

           0       0.85      0.96      0.90     12435
           1       0.77      0.45      0.57      3846

    accuracy                           0.84     16281
   macro avg       0.81      0.70      0.73     16281
weighted avg       0.83      0.84      0.82     16281



In [340]:
# MARITAL STATUS

# Creating training and testing datasets with 'marital status' added
x_marital = pd.concat([x[num_columns], x[marital]], axis=1)
xt_marital = pd.concat([xt[num_columns], xt[marital]], axis=1)

# Separating features from target
y_marital = y.copy()
yt_marital = yt.copy()

# Initialize and train Random Forest Classifier
rf_model = RandomForestClassifier(criterion='entropy', max_depth=15, random_state=42)
rf_model.fit(x_marital, y_marital)

# Predict on test set
rf_predictions = rf_model.predict(xt_marital)

# Evaluate model performance
print("Random Forest Classifier (Numerical Features + 'Marital Status')")
print("Accuracy:", accuracy_score(yt_marital, rf_predictions))
print("Confusion Matrix:\n", confusion_matrix(yt_marital, rf_predictions))
print("Classification Report:\n", classification_report(yt_marital, rf_predictions))

Random Forest Classifier (Numerical Features + 'Marital Status')
Accuracy: 0.8628462625145875
Confusion Matrix:
 [[11834   601]
 [ 1632  2214]]
Classification Report:
               precision    recall  f1-score   support

           0       0.88      0.95      0.91     12435
           1       0.79      0.58      0.66      3846

    accuracy                           0.86     16281
   macro avg       0.83      0.76      0.79     16281
weighted avg       0.86      0.86      0.85     16281



In [341]:
# RACE

# Creating training and testing datasets with 'race' added
x_race = pd.concat([x[num_columns], x[race]], axis=1)
xt_race = pd.concat([xt[num_columns], xt[race]], axis=1)

# Separating features from target
y_race = y.copy()
yt_race = yt.copy()

# Initialize and train Random Forest Classifier
rf_model = RandomForestClassifier(criterion='entropy', max_depth=15, random_state=42)
rf_model.fit(x_race, y_race)

# Predict on test set
rf_predictions = rf_model.predict(xt_race)

# Evaluate model performance
print("Random Forest Classifier (Numerical Features + 'Race')")
print("Accuracy:", accuracy_score(yt_race, rf_predictions))
print("Confusion Matrix:\n", confusion_matrix(yt_race, rf_predictions))
print("Classification Report:\n", classification_report(yt_race, rf_predictions))

Random Forest Classifier (Numerical Features + 'Race')
Accuracy: 0.8384620109329893
Confusion Matrix:
 [[11932   503]
 [ 2127  1719]]
Classification Report:
               precision    recall  f1-score   support

           0       0.85      0.96      0.90     12435
           1       0.77      0.45      0.57      3846

    accuracy                           0.84     16281
   macro avg       0.81      0.70      0.73     16281
weighted avg       0.83      0.84      0.82     16281



In [342]:
# SEX

# Creating training and testing datasets with 'sex' added
x_sex = pd.concat([x[num_columns], x[sex]], axis=1)
xt_sex = pd.concat([xt[num_columns], xt[sex]], axis=1)

# Separating features from target
y_sex = y.copy()
yt_sex = yt.copy()

# Initialize and train Random Forest Classifier
rf_model = RandomForestClassifier(criterion='entropy', max_depth=15, random_state=42)
rf_model.fit(x_sex, y_sex)

# Predict on test set
rf_predictions = rf_model.predict(xt_sex)

# Evaluate model performance
print("Random Forest Classifier (Numerical Features + 'sex')")
print("Accuracy:", accuracy_score(yt_sex, rf_predictions))
print("Confusion Matrix:\n", confusion_matrix(yt_sex, rf_predictions))
print("Classification Report:\n", classification_report(yt_sex, rf_predictions))

Random Forest Classifier (Numerical Features + 'sex')
Accuracy: 0.8420858669614889
Confusion Matrix:
 [[11875   560]
 [ 2011  1835]]
Classification Report:
               precision    recall  f1-score   support

           0       0.86      0.95      0.90     12435
           1       0.77      0.48      0.59      3846

    accuracy                           0.84     16281
   macro avg       0.81      0.72      0.75     16281
weighted avg       0.83      0.84      0.83     16281



**Interpretation:** *Note: It was not specified how many additional variables to test, so I chose to exclude those that were not already accounted for by numeric variables (e.g., years of education) or with an excessive number of options (e.g., native country, occupation). If this was incorrect, please let me know and I will resubmit the assignmnet, however initial instructions were not clear.*

In this analysis, I systematically introduced new, non-numerical columns (i.e., workclass, marital status, sex, and race) one by one into a Random Forest Classifier trained on the adult dataset. Each additional non-numerical feature introduced varying degrees of impact on the Random Forest Classifier's performance. Notably, marital status appeared to be the most beneficial, increasing the accuracy rate from ~84% with just the numerical variables to over 86% with marital status added. This increase in accuracy suggests that marital status holds significant predictive power regarding income levels. Additionally, the model including marital status performed better in terms of precision and recall compared to the numerical values only model. 
Conversely, workclass, race, and sex showed minimal impact on model perofmrance with minute changes in accuracy, precision, recall, and F1 scores between their individual models and the baseline numerical model. 